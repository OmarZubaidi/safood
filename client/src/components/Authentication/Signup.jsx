// Package imports
import React, { useState, useRef } from 'react';
import { Link, useNavigate } from 'react-router-dom';
import { Card, Button, Form, Alert } from 'react-bootstrap';

// Package imports
import { useAuth } from '../../context/AuthContext';
import { postUser } from '../service';

export default function Signup () {
  // Refs, states, navigation, and authentication
  const emailRef = useRef();
  const passwordRef = useRef();
  const passwordConfirmRef = useRef();
  const nameRef = useRef();
  const aboutRef = useRef();
  const [error, setError] = useState();
  const [loading, setLoading] = useState(false);
  const navigate = useNavigate();
  const { signup } = useAuth();

  // Submit function
  async function handleSubmit (e) {
    e.preventDefault();

    if (passwordRef.current.value !== passwordConfirmRef.current.value) {
      return setError("Passwords don't match.");
    }

    try {
      setError('');
      setLoading(true);
      const auth = await signup(
        emailRef.current.value,
        passwordRef.current.value
      );
      await postUser({
        name: nameRef.current.value,
        events: [],
        allergens: [],
        uid: auth.user.uid,
        about: aboutRef.current.value
          || `Hi, my name is ${nameRef.current.value}`,
        img: ''
      });
      navigate('/');
    } catch (error) {
      console.log(error);
      setError('Failed to create an account.');
    }
    setLoading(false);
  }

  return (
    <>
      <Card>
        <Card.Body>
          <h2 className='text-center mb-4'>
            Sign Up
          </h2>
          {error && <Alert variant='danger'>
            {error}
          </Alert>}
          <Form onSubmit={handleSubmit}>
            <Form.Group id='name'>
              <Form.Label>
                Name
              </Form.Label>
              <Form.Control
                type='text'
                ref={nameRef}
                required
              />
            </Form.Group>
            <Form.Group id='email'>
              <Form.Label>
                E-mail
              </Form.Label>
              <Form.Control
                type='email'
                ref={emailRef}
                required
              />
            </Form.Group>
            <Form.Group id='password'>
              <Form.Label>
                Password
              </Form.Label>
              <Form.Control
                type='password'
                ref={passwordRef}
                required
              />
            </Form.Group>
            <Form.Group id='passwordconfirm'>
              <Form.Label>
                Confirm password
              </Form.Label>
              <Form.Control
                type='password'
                ref={passwordConfirmRef}
                required
              />
            </Form.Group>
            <Form.Group id='about'>
              <Form.Label>
                Tell us something about yourself
              </Form.Label>
              <Form.Control
                type='text'
                ref={aboutRef}
              />
            </Form.Group>
            <Button
              disabled={loading}
              className='w-100'
              type='submit'
            >
              Sign Up
            </Button>
          </Form>
        </Card.Body>
      </Card>
      <div className='w-100 text-center mt-2'>
        Already have an account? <Link to='/login'>
          Sign in
        </Link>
      </div>
    </>
  );
}
