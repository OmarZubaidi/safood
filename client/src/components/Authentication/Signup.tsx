// Package imports
import React, { useState, useRef } from 'react';
import { Link, useNavigate } from 'react-router-dom';
import { Card, Button, Form, Alert } from 'react-bootstrap';

// Package imports
import { useAuth } from '../../context/AuthContext';
import { postUser } from '../../services';

export default function Signup () {
  // Refs, states, navigation, and authentication
  const emailRef = useRef<HTMLInputElement>(null);
  const passwordRef = useRef<HTMLInputElement>(null);
  const passwordConfirmRef = useRef<HTMLInputElement>(null);
  const nameRef = useRef<HTMLInputElement>(null);
  const aboutRef = useRef<HTMLInputElement>(null);
  const [error, setError] = useState<string>();
  const [loading, setLoading] = useState<boolean>(false);
  const navigate = useNavigate();
  const { signup } = useAuth();

  // Submit function
  async function handleSubmit (e: React.FormEvent<HTMLFormElement>) {
    e.preventDefault();

    if (passwordRef.current?.value !== passwordConfirmRef.current?.value) {
      return setError("Passwords don't match.");
    }

    try {
      setError('');
      setLoading(true);

      const name = nameRef.current?.value;
      const email = emailRef.current?.value;
      const password = passwordRef.current?.value;
      const about = aboutRef.current?.value || `Hi, my name is ${name}`;
      if(!email || !password || !name) {
        setError('Email, password, name are required!');
        return setLoading(false);
      }

      const auth = await signup(
        email,
        password
      );

      await postUser({
        name,
        events: [],
        allergens: [],
        uid: auth.user.uid,
        about: about,
        img: ''
      });
      navigate('/');
    } catch (error) {
      console.error(error);
      setError('Failed to create an account.');
    }
    setLoading(false);
  }

  return (
    <>
      <Card>
        <Card.Body>
          <h2 className='text-center mb-4'>
            Sign Up
          </h2>
          {error && <Alert variant='danger'>
            {error}
          </Alert>}
          <Form onSubmit={handleSubmit}>
            <Form.Group id='name'>
              <Form.Label
                htmlFor="name-input"
              >
                Name
              </Form.Label>
              <Form.Control
                type='text'
                ref={nameRef}
                id="name-input"
                required
              />
            </Form.Group>
            <Form.Group id='email'>
              <Form.Label
              htmlFor="email-input">
                E-mail
              </Form.Label>
              <Form.Control
                type='email'
                ref={emailRef}
                id="email-input"
                required
              />
            </Form.Group>
            <Form.Group id='password'>
              <Form.Label
              htmlFor="password-input">
                Password
              </Form.Label>
              <Form.Control
                type='password'
                ref={passwordRef}
                id="password-input"
                required
              />
            </Form.Group>
            <Form.Group id='passwordconfirm'>
              <Form.Label
                htmlFor="passwordconfirm-input"
              >
                Confirm password
              </Form.Label>
              <Form.Control
                type='password'
                id="passwordconfirm-input"
                ref={passwordConfirmRef}
                required
              />
            </Form.Group>
            <Form.Group id='about'>
              <Form.Label
              htmlFor="about-input">
                Tell us something about yourself
              </Form.Label>
              <Form.Control
                type='text'
                ref={aboutRef}
                id="about-input"
              />
            </Form.Group>
            <Button
              disabled={loading}
              className='w-100'
              type='submit'
            >
              Sign Up
            </Button>
          </Form>
        </Card.Body>
      </Card>
      <div className='w-100 text-center mt-2'>
        Already have an account? <Link to='/login'>
          Sign in
        </Link>
      </div>
    </>
  );
}
