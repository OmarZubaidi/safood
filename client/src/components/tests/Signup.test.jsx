// Package imports
import { render, screen } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';

// Local imports
import Signup from '../Authentication/Signup';

const mockLogin = {
  name: "nick",
  email: "n@1.com",
  password: '123456'

};

jest.mock('../../context/AuthContext.js', () => ({
  useAuth: () => {
    return {
      signup: (email, password ) => {
      return {
        user: {
          uid: '12345'
        }
      }
    }
  } 
  }
}));

jest.mock('../../services', () => ({
  postUser: (user) => {
    console.log(user);
  } 
}));

describe('Signup component', () => {

  test('should match the snapshot', () => {
    const { container } = render(
      <BrowserRouter>
        <Signup />
      </BrowserRouter>
    );
    expect(container).toMatchSnapshot();
  });

  test('should render the signup component', () => {
    render(
      <BrowserRouter>
        <Signup />
      </BrowserRouter>
    );
    screen.getByText('Name');
    screen.getByText('E-mail');
    screen.getByText('Password');
    screen.getByText('Confirm password');
    screen.getByText('Tell us something about yourself');
    screen.getByRole('button');
  });

});