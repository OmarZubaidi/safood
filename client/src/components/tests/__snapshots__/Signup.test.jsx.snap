// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Signup component should match the snapshot 1`] = `
<div>
  <div
    class="card"
  >
    <div
      class="card-body"
    >
      <h2
        class="text-center mb-4"
      >
        Sign Up
      </h2>
      <form
        class=""
      >
        <div
          id="name"
        >
          <label
            class="form-label"
          >
            Name
          </label>
          <input
            class="form-control"
            required=""
            type="text"
          />
        </div>
        <div
          id="email"
        >
          <label
            class="form-label"
          >
            E-mail
          </label>
          <input
            class="form-control"
            required=""
            type="email"
          />
        </div>
        <div
          id="password"
        >
          <label
            class="form-label"
          >
            Password
          </label>
          <input
            class="form-control"
            required=""
            type="password"
          />
        </div>
        <div
          id="passwordconfirm"
        >
          <label
            class="form-label"
          >
            Confirm password
          </label>
          <input
            class="form-control"
            required=""
            type="password"
          />
        </div>
        <div
          id="about"
        >
          <label
            class="form-label"
          >
            Tell us something about yourself
          </label>
          <input
            class="form-control"
            type="text"
          />
        </div>
        <button
          class="w-100 btn btn-primary"
          type="submit"
        >
          Sign Up
        </button>
      </form>
    </div>
  </div>
  <div
    class="w-100 text-center mt-2"
  >
    Already have an account? 
    <a
      href="/login"
    >
      Sign in
    </a>
  </div>
</div>
`;
